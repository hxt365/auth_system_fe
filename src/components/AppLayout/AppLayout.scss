.message-result {
  padding: 5rem 3rem 3rem 3rem !important;
}
