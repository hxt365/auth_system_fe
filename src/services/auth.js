// @flow

import {
  LOGIN,
  LOGOUT,
  REFRESH,
  RESET_PASSWORD,
  CHANGE_PASSWORD,
  SIGNUP,
  CONFIRM_EMAIL,
} from 'constants/api';

import type {
  loginFormType,
  resetPasswordType,
  changePasswordFormType,
  signupFormType,
} from 'type';

import customAxios from './axios';

const login = async (data: loginFormType) => {
  const res = await customAxios.post(LOGIN, data).catch(err => err);
  return res;
};

const logout = async () => {
  const res = await customAxios.post(LOGOUT).catch(err => err);
  return res;
};

const resetPassword = async (data: resetPasswordType) => {
  const res = await customAxios.post(RESET_PASSWORD, data).catch(err => err);
  return res;
};

const refresh = async () => {
  const res = await customAxios.post(REFRESH).catch(err => err);
  return res;
};

const changePassword = async (data: changePasswordFormType) => {
  const res = await customAxios.post(CHANGE_PASSWORD, data).catch(err => err);
  return res;
};

const signup = async (data: signupFormType) => {
  const res = await customAxios.post(SIGNUP, data).catch(err => err);
  return res;
};

const confirmEmail = async (token: String) => {
  const link = `${CONFIRM_EMAIL + token}/`;
  const res = await customAxios.post(link).catch(err => err);
  return res;
};

export default {
  login,
  logout,
  refresh,
  resetPassword,
  changePassword,
  signup,
  confirmEmail,
};
