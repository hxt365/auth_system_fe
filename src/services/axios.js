// @flow

import axios from 'axios';

export const BASE_API_URL: String = `${process.env.REACT_APP_BASE_API_URL}`;

const customAxios = axios.create({
  baseURL: BASE_API_URL,
  withCredentials: true,
  xsrfCookieName: 'csrftoken',
  xsrfHeaderName: 'X-CSRFToken',
});

customAxios.interceptors.response.use(
  response => response,
  error => {
    // Add logic to handle error here
    console.log(error.response);
    return Promise.reject(error);
  },
);

export default customAxios;
