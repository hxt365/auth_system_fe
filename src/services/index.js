export { default as authServices } from './auth';
