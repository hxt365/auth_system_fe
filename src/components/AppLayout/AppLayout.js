// @flow

import React, { useState, useEffect } from 'react';
import { Modal, Result, Button } from 'antd';
import './AppLayout.scss';
import { SUCCESS, ERROR } from 'constants/message';
import { withRouter } from 'react-router-dom';
import { authServices } from 'services';
import { LOGIN } from 'constants/route';

const COUNTDOWN = parseInt(process.env.REACT_APP_COUNTDOWN, 10);

type PropsType = {
  children: any,
  history: any,
};

// eslint-disable-next-line no-unused-vars
type MessageType = {
  status: SUCCESS | ERROR,
  title: String,
  button: String,
  redirect: Boolean,
};

// eslint-disable-next-line no-unused-vars
type UserType = {
  username: String,
  first_name: String,
  last_name: String,
  is_authenticated: String,
};

export const AppContext = React.createContext(null);

const initialMessage = {
  status: '',
  title: '',
  button: '',
  redirect: '',
};

const initialUser = {
  username: '',
  first_name: '',
  last_name: '',
  is_authenticated: '',
};

function AppLayout(props: PropsType) {
  const [message: MessageType, setMessage] = useState(initialMessage);

  const [user: UserType, setUser] = useState(initialUser);

  useEffect(() => {
    (async function IFE() {
      const res = await authServices.refresh();
      if (res.status === 200) {
        setUser({
          username: res.data.username,
          first_name: res.data.first_name,
          last_name: res.data.last_name,
          is_authenticated: true,
        });
      }
    })();
  }, []);

  useEffect(() => {
    // Auto logout user
    setTimeout(() => {
      if (user.is_authenticated) setUser(initialUser);
    }, COUNTDOWN);
  }, [user]);

  const { children } = props;

  const clickHandler = e => {
    e.preventDefault();
    if (message.redirect === LOGIN) setUser(initialUser);
    if (message.redirect) props.history.push(message.redirect);
    setMessage(initialMessage);
  };

  return (
    <AppContext.Provider value={{ user, setUser, setMessage }}>
      {message.status && message.title ? (
        <Modal visible centered closable={false} footer={null} bodyStyle={{ padding: '1rem' }}>
          <Result
            status={message.status}
            title={message.title}
            className="message-result"
            extra={[
              <Button key="btn" type="primary" onClick={clickHandler}>
                {message.button}
              </Button>,
            ]}
          />
        </Modal>
      ) : null}
      {children}
    </AppContext.Provider>
  );
}

export default withRouter(AppLayout);
