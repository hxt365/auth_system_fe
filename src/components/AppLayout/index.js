import AppLayout, { AppContext } from './AppLayout';

export default AppLayout;
export { AppContext };
