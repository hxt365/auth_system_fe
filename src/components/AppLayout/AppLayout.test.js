import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import { authServices } from 'services';
import { BrowserRouter } from 'react-router-dom';
import AppLayout from './AppLayout';

describe('test AppLayout', () => {
  it('test snapshot', () => {
    const wrapper = shallow(<AppLayout />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('make sure refresh called once', () => {
    const refresh = jest.spyOn(authServices, 'refresh').mockImplementation(() => {});
    const wrapper = mount(
      <BrowserRouter>
        <AppLayout />
      </BrowserRouter>,
    );
    expect(refresh).toBeCalledTimes(1);
    refresh.mockClear();
    wrapper.unmount();
  });
});
