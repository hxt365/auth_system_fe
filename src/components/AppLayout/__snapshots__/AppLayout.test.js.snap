// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AppLayout test snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
